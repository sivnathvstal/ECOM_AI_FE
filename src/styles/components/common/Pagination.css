.pagination-container {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 16px 0;
}

.pagination-info {
  margin: 0 16px;
} 