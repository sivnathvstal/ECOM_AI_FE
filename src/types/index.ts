export interface Product {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
  rating: {
    rate: number;
    count: number;
  };
}

export interface CartItem {
  id: number;
  productId: number;
  quantity: number;
  product: Product;
}

export interface Cart {
  id: number;
  userId: number;
  date: string;
  products: CartProduct[];
}

export interface CartProduct {
  productId: number;
  quantity: number;
  product?: Product;
}

export interface PaginationParams {
  page: number;
  limit: number;
  total: number;
}

export interface ProductsResponse {
  products: Product[];
  pagination: PaginationParams;
} 