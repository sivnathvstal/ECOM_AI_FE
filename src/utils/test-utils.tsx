import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider, createTheme } from '@material-ui/core';
import { QueryClient, QueryClientProvider } from 'react-query';
import { CartProvider } from '../context/CartContext';

// Create a theme for testing
const theme = createTheme({
  palette: {
    primary: {
      main: '#3f51b5',
    },
    secondary: {
      main: '#f50057',
    },
  },
});

// Create a client for React Query
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

// All providers wrapper
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider theme={theme}>
        <CartProvider>
          <BrowserRouter>{children}</BrowserRouter>
        </CartProvider>
      </ThemeProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// Re-export everything
export * from '@testing-library/react';

// Override render method
export { customRender as render }; 