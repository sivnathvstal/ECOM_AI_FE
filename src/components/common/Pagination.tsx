import React from 'react';
import { 
  Box, 
  Button
} from '@material-ui/core';
import { 
  NavigateBefore, 
  NavigateNext 
} from '@material-ui/icons';
import { PaginationParams } from '../../types';
import './Pagination.css';

interface PaginationProps {
  pagination: PaginationParams;
  onPageChange: (page: number) => void;
}

const Pagination: React.FC<PaginationProps> = ({ pagination, onPageChange }) => {
  const { page, limit, total } = pagination;
  
  const totalPages = Math.ceil(total / limit);
  const hasPrevious = page > 1;
  const hasNext = page < totalPages;
  
  const handlePrevious = () => {
    if (hasPrevious) {
      onPageChange(page - 1);
    }
  };
  
  const handleNext = () => {
    if (hasNext) {
      onPageChange(page + 1);
    }
  };
  
  return (
    <Box className="pagination-container" data-testid="pagination">
      <Button
        color="primary"
        disabled={!hasPrevious}
        onClick={handlePrevious}
        startIcon={<NavigateBefore />}
        data-testid="previous-page"
      >
        Previous
      </Button>
      <Box className="pagination-info" data-testid="page-info">
        Page {page} of {totalPages}
      </Box>
      <Button
        color="primary"
        disabled={!hasNext}
        onClick={handleNext}
        endIcon={<NavigateNext />}
        data-testid="next-page"
      >
        Next
      </Button>
    </Box>
  );
};

export default Pagination; 