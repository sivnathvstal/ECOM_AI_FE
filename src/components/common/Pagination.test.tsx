import React from 'react';
import { render, screen, fireEvent } from '../../utils/test-utils';
import Pagination from './Pagination';
import { PaginationParams } from '../../types';

describe('Pagination', () => {
  const mockOnPageChange = jest.fn();
  
  const renderPagination = (pagination: PaginationParams) => {
    return render(
      <Pagination 
        pagination={pagination}
        onPageChange={mockOnPageChange}
      />
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { container } = renderPagination({ page: 1, limit: 8, total: 24 });
    expect(container).toMatchSnapshot();
  });

  it('displays current page and total pages', () => {
    renderPagination({ page: 2, limit: 8, total: 24 });
    expect(screen.getByTestId('page-info')).toHaveTextContent('Page 2 of 3');
  });

  it('disables previous button on first page', () => {
    renderPagination({ page: 1, limit: 8, total: 24 });
    const prevButton = screen.getByTestId('previous-page');
    expect(prevButton).toBeDisabled();
  });

  it('enables previous button when not on first page', () => {
    renderPagination({ page: 2, limit: 8, total: 24 });
    const prevButton = screen.getByTestId('previous-page');
    expect(prevButton).not.toBeDisabled();
  });

  it('disables next button on last page', () => {
    renderPagination({ page: 3, limit: 8, total: 24 });
    const nextButton = screen.getByTestId('next-page');
    expect(nextButton).toBeDisabled();
  });

  it('enables next button when not on last page', () => {
    renderPagination({ page: 2, limit: 8, total: 24 });
    const nextButton = screen.getByTestId('next-page');
    expect(nextButton).not.toBeDisabled();
  });

  it('calls onPageChange with previous page when previous button is clicked', () => {
    renderPagination({ page: 2, limit: 8, total: 24 });
    const prevButton = screen.getByTestId('previous-page');
    fireEvent.click(prevButton);
    expect(mockOnPageChange).toHaveBeenCalledWith(1);
  });

  it('calls onPageChange with next page when next button is clicked', () => {
    renderPagination({ page: 2, limit: 8, total: 24 });
    const nextButton = screen.getByTestId('next-page');
    fireEvent.click(nextButton);
    expect(mockOnPageChange).toHaveBeenCalledWith(3);
  });

  it('does not call onPageChange when previous button is clicked on first page', () => {
    renderPagination({ page: 1, limit: 8, total: 24 });
    const prevButton = screen.getByTestId('previous-page');
    fireEvent.click(prevButton);
    expect(mockOnPageChange).not.toHaveBeenCalled();
  });

  it('does not call onPageChange when next button is clicked on last page', () => {
    renderPagination({ page: 3, limit: 8, total: 24 });
    const nextButton = screen.getByTestId('next-page');
    fireEvent.click(nextButton);
    expect(mockOnPageChange).not.toHaveBeenCalled();
  });
}); 