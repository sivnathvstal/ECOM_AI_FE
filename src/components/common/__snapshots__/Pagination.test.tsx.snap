// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-1 pagination-container"
    data-testid="pagination"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      data-testid="previous-page"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
        </span>
        Previous
      </span>
    </button>
    <div
      class="MuiBox-root MuiBox-root-2 pagination-info"
      data-testid="page-info"
    >
      Page 
      1
       of 
      3
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      data-testid="next-page"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Next
        <span
          class="MuiButton-endIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
